src/ast/ast_validations.c

```diff
@@ -1409,6 +1400,83 @@ static AST_Validation _ValidateClauses(const AST *ast, char **reason) {
 	return AST_VALID;
 }
 
+static AST_Validation _ValidateUnion_Clauses(const AST *ast, char **reason) {
+	if(!AST_ContainsClause(ast, CYPHER_AST_UNION)) return AST_VALID;
+
+	/* Make sure there's no conflict between UNION clauses
+	 * either all UNION clauses specify ALL or nither of them does. */
+	AST_Validation res = AST_VALID;
+	uint *union_indices = AST_GetClauseIndices(ast, CYPHER_AST_UNION);
+	uint union_clause_count = array_len(union_indices);
+	int has_all_count = 0;
+
+	for(uint i = 0; i < union_clause_count; i++) {
+		const cypher_astnode_t *union_clause = cypher_ast_query_get_clause(ast->root, union_indices[i]);
+		if(cypher_ast_union_has_all(union_clause)) has_all_count++;
+	}
+	array_free(union_indices);
+
+	// If we've encountered UNION ALL clause, all UNION clauses should specify ALL.
+	if(has_all_count != 0) {
+		if(has_all_count != union_clause_count) {
+			asprintf(reason, "Invalid combination of UNION and UNION ALL.");
+			return AST_INVALID;
+		}
+	}
+
+	// Require all RETURN clauses to perform the exact same projection.
+	uint *return_indices = AST_GetClauseIndices(ast, CYPHER_AST_RETURN);
+	uint return_clause_count = array_len(return_indices);
+
+	const cypher_astnode_t *return_clause = cypher_ast_query_get_clause(ast->root, return_indices[0]);
+	uint proj_count = cypher_ast_return_nprojections(return_clause);
+	const char *projections[proj_count];
+
+	for(uint j = 0; j < proj_count; j++) {
+		const cypher_astnode_t *proj = cypher_ast_return_get_projection(return_clause, j);
+		const cypher_astnode_t *alias_node = cypher_ast_projection_get_alias(proj);
+		if(alias_node == NULL)  {
+			projections[j] = NULL;
```

Comment:
In which cases projection will have NULL alias?

Reply:
`RETURN a`
VS
`RETURN a AS b`